### ISSUE TITLE
2.0.3: issue with `Requires.private`

### ISSUE THREAD

**Issue by kloczek** [pkgconf/pkgconf#321]
## Preface
From pkgconfig pc(5) man page:
```man
     Requires.private
             Required dependencies that must be met for the package to be usable for static linking. 
             All dependencies must be satisfied or the pkg-config implementation must not use
             the package for >>>static linking<<<.  (optional; dependency list)
```
## The case
On building building dynamically linked xsetroot pkgconfig fails because it as not able to find in build env Requires.private dependencies listed in xcursor.pc file:
```console
checking for xmuu x11 xbitmaps xcursor xproto >= 7.0.25... no
configure: error: Package requirements (xmuu x11 xbitmaps xcursor xproto >= 7.0.25) were not met:

Package 'xrender', required by 'xcursor', not found
Package 'xfixes', required by 'xcursor', not found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables XSETROOT_CFLAGS
and XSETROOT_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
```
From config.og:
```console
configure:5502: checking for xmuu x11 xbitmaps xcursor xproto >= 7.0.25
configure:5509: $PKG_CONFIG --exists --print-errors "xmuu x11 xbitmaps xcursor xproto >= 7.0.25"
configure:5512: $? = 0
configure:5526: $PKG_CONFIG --exists --print-errors "xmuu x11 xbitmaps xcursor xproto >= 7.0.25"
configure:5529: $? = 0
configure:5543: result: no
Package 'xrender', required by 'xcursor', not found
Package 'xfixes', required by 'xcursor', not found
configure:5559: error: Package requirements (xmuu x11 xbitmaps xcursor xproto >= 7.0.25) were not met:

Package 'xrender', required by 'xcursor', not found
Package 'xfixes', required by 'xcursor', not found
```
So in this case pkgconfig checks not only Requres: but Requires.privte:
```pkgconfig
Content of the xcursor.pc:
prefix=/usr
exec_prefix=/usr
libdir=/usr/lib64
includedir=/usr/include
datarootdir=${prefix}/share
icondir=/usr/share/icons

Name: Xcursor
Description: X Cursor Library
Version: 1.2.1
Requires: xproto
Requires.private: x11 xrender xfixes
Cflags: -I${includedir}
Libs: -L${libdir} -lXcursor
```
`xcursor` header files are using ONLY header files provides by `xproto`.
## Conclutions:
- pkgconfig 1.x did not processed in such cases `Requires.private` and it was OK because by default all GBU autotools, meson and cmake based buil descriton which are usimg pkgconfig by default are used to build dynamic linking. Static linking can be forced by use env variable passed to those build framework like `LDFLAGS=-static` however none of those build frameworks changes pakconfig execution params depending on that env variable,
- If it is some other method which allows skip Requires.private dependencies pkg.m4 aclocal macros should be adjusted (before start build I'm always regenerating all GNU autotools files by execute `autorecoobf -fiv`)
- In this case I can apply workaround on packaging libXcursor by remove `Requires.private` (and it could make sense as my libXcursor -devel package does not provide statis libXcursor.a) from provided by this package xcursor.pc. However it will be much more such cases.
- Because all above and because I see similar issues in other projects which are using cmake and meson IMO it is bug in pkgconfig because as it is literally described in pc(5) man page Requires.private (and other *.private) should be processed ONLY when STATIC linking is used.

**Comment by kaniini** [comment]
No change has been made to the behavior of `Requires.private`, it has always been checked when cflags are requested due to the possibility that public header files may reference other header files.

**Comment by kloczek** [comment]
> No change has been made to the behavior of Requires.private, it has always been checked when cflags are requested due to the possibility that public header files may reference other header files.

- Compare to version 1.9.5 this is completely different behaviour
- Just closing ticket with note that "it has always been checked" is not king solation in context of mentioned/described issue.

**Comment by dg0yt** [comment]
Coming back here while studying test cases. In particular `tests/lib1/metapackage-1.pc`.

- Maybe it was added in 1.9.3? https://github.com/pkgconf/pkgconf/commit/746005582fbc0719de49310d2a836ae379da7792

- > it has always been checked when cflags are requested due to the possibility that public header files may reference other header files.

  If there is a public dependency, it belongs into `Requires` IMO.
  I didn't expect to be required to install any modules from private lists unless using `--static`.

- But I see that pkg-config 0.29.1 behaves the same.

This behavior is scary! You may have a different solution for `--cflags` than for `--libs`. Maybe no solution at all with `Requires.private`. Maybe a binary incompatible solution (`-DWITH_FOO -lwithout_foo`). 

If consistency is desired, and `--cflags` is meant to consider `Requires.private`, than the solution always has to consider `Requires.private` for determining the initial solution, and omit output for the packages which are only reachable via  `Requires.private` when they are not to be considered (`--libs`).

**Comment by kloczek** [comment]
FYI: I found workaround for that issue.
Because I'm not building any packages with static libraries I've patched all my packages wo add remove all `Requires.private:`, `Cflags.private` and `Libs.private`.
As long as there is those fields in .pc files all works correctly. because in rpm is IMO another bug related to way how rpm pkgconfig dependencies are generated https://github.com/rpm-software-management/rpm/blob/master/scripts/pkgconfigdeps.sh#L42-L60 which is causing that to generated rpm `PROVIDES` are added all `Requires.private` (effectively from point of view of that scrips there is no difference between `Requires:` and `Requires.private`) it caused that in few rare cases on remove `Requires.private:` it is necessary to add some `Requires:` entries.
In this case problematic is use `--print-requires-private` with `--print-requires`.

I have already +4.8k spec files of my packages distro and ..
```console
[tkloczko@pers-jacek SPECS]$ grep "Patch:.*%{name}-no_pkgconfig_private.patch" *spec | wc -l
122
```

IMO meson should not at all generate .pc files with `Requires.private:`, `Cflags.private` and `Libs.private` fields if it is used to build **only shared libraries**. I've opened ticket for for that https://github.com/mesonbuild/meson/issues/12225 but looks like even here is complete lack of understanding of that issue 😞 
I have my own patch on meson which skips generate any Requires.private:`, `Cflags.private` and `Libs.private` fields.

In other words: IMO root cause of this issue is pkgconfigdeps.sh included in rpm which caused that issues with use `Requires.private:` the same way as `Requires:` and this should be fixed first. Than meson ad pkgconfig should be adjusted as well.

As I wrote it is workaround solution to patch all source code of those projects which have statically generated its own .pc files (from some .in files) + meson fix to skip use those fields in `pc.generate()` meson calls (my patch removes taht uncoinditionally but it should be extended only when meson `pc.generate()` is used to build shared only version of some library.
It is only kind workaround however those issues shuld be IMO addressed in ALL THHREE projects together because if not it will cause a lot of disputes/chaos/waste of time.

I've already wasted enough time on that issue and have no time convince at least three projects maintainers (rpm, meson and pkgconfig) to solve that issue in one go.
Feel free to use refs from this ticket to try convince more people that issue IS REAL and partially it is BrokenByDesign™️ issue.

**Comment by dg0yt** [comment]
There is a fundamental issue with the design of `.private` but that's out of scope for me. I can accept the `--cflags` behavior if it is the de-facto standard.

My point is that `pkgconf` should not construct different dependency resolution problems for `--cflags` vs. `--libs`.
